design/velero-client-download-apiserver_design.md

```diff
@@ -0,0 +1,128 @@
+Even if the object store is accessible from the client side, today things like cacert and insecure-skip-tls-verify flags are specified manually from the client even though the Velero server already has the information.
+
+This design allows velero CLI to request files from any object store plugin that velero server can call GetObject() without creating a signedUrl first.
+## Goals
+- Enable Velero CLI users to download data from the Velero server without having to create signedUrl for scenarios where the storage provider is not able to create signedUrl accessible from the velero CLI client side.
+
+## Non Goals
+- Making the download server the default way to download data from object store.
+
+## High-Level Design
+Create an endpoint on the Velero server that CLI users can connect to and download data from without requiring signed URL from object store directly when configured explicitly or as a fallback.
+
+### Velero Download Server URL generation using Service to Expose Download Server
+
+Upon enabling Download Server API, Velero will create a service of type that user specifies, which will expose the download server to the outside world. The download server will proxy requested data from the object store and stream it back to the client using one of the available service types (ClusterIP/NodePort/LoadBalancer/ExternalName). User can also precreate a service and pass the service name to velero during install time with `velero install --veleroServerServiceName=veleroDownloadServerServiceName` to use the precreated service to expose the download server.
+
+If the service can also be reached via a fully qualified domain name (FQDN), user can pass the FQDN to velero during install time with `velero install --veleroServerFQDN=dnsNameToVeleroServerService.com`. Velero Download Server will use FQDN in-place of values auto discovered from created service (ie. NodePort/Cluster IP)
+
+This approach may also [work on KinD clusters](https://kind.sigs.k8s.io/docs/user/loadbalancer/)
+
+The current DownloadRequest CR status `veleroServerURL` field will be populated to use the Download Server URL instead of the signed object store URL where available.
+
+Velero will be updated to download data from the object store and stream it back to the client.
+
+The Velero CLI will still default to `downloadURL` from DownloadRequest but can fallback or forced to `veleroServerURL` which is served via the Velero server.
+
+If velero CLI attempted request from veleroServerURL at the specified FQDN and failed, and velero CLI command was executed inside velero server pod (`alias velero='kubectl exec -it deployment/velero -- velero'`), velero CLI will attempt to use the service name (cluster.local address) to connect to the download server.
+
+[![Velero Download Server](VeleroDownloadServer.jpg)](VeleroDownloadServer.jpg)
+
+## Detailed Design
```

Comment:
Could you add a section about what the endpoint looks like?

Reply:
Yes. Will do. I think we can simply use the same path pattern as to how the files are stored today.. so something like `/backups/<backupName>/<contents>/...`